[cli-tcp] add TLS mode to TCP CLI tool and TLS callbacks to TCP API

samkumar opened the pull request:
This pull request adds support for TLS in the TCP CLI tool, along with a test of TLS functionality.

The main purpose of this is to provide an example of how to use mbedTLS  with TCPlp to run TLS. I felt it is important to have an example because I expect that many applications that use TCP, particularly those in which a Thread device exchanges data with a device outside the Thread mesh, will need to run TLS over TCP for security reasons.

A secondary benefit is to add TCP functionality to the TCP CLI tool. Additionally, by having the example code be part of a "real" application (the CLI tool), there is less risk of the code going stale as OpenThread evolves.

I added an "expects" test for the TLS functionality in the CLI tool. I also ran the code on two Nordic NRF52840-DK boards to make sure it works. I've included a [ZIP of a packet capture](https://github.com/openthread/openthread/files/9740767/tls-example-capture.zip) of TLS in OpenThread, collected using a packet sniffer when using the TCP CLI tool to exchange data between the two Nordic NRF52840-DK devices. (The Network Key is `45c96ebaa174bdb955c0151db892d8ca` and the Mesh Local Prefix is `fd5f:ad2d:4d98:363b::/64`.) The workload is based on the first TCP connection in the "expects" test.

The performance impact of TLS appears to be small. When using TCP without TLS, I am able to achieve ~80 kb/s goodput between two NRF52840-DK boards. With TLS in the same setup, I am able to achieve ~70-75 kb/s goodput.

edmont commented on src/cli/cli_tcp.hpp:
```diff
@@ -39,6 +39,11 @@
 #include <openthread/tcp.h>
 #include <openthread/tcp_ext.h>
 
+#include <mbedtls/certs.h>
```
See https://github.com/Mbed-TLS/mbedtls/blob/development/docs/3.0-migration-guide.md#remove-the-certs-module-from-the-library

edmont replied:
The include seems to be unneeded.

samkumar replied:
Thanks for letting me know.

samkumar replied:
I've removed it in my next commit.

edmont commented on src/cli/cli_tcp.cpp:
```diff
@@ -74,13 +98,62 @@ template <> otError TcpExample::Process<Cmd("init")>(Arg aArgs[])
+            mbedtls_ssl_conf_authmode(&mSSLConfig, MBEDTLS_SSL_VERIFY_NONE);
+            mbedtls_entropy_init(&mEntropy);
+
+            int rv = mbedtls_pk_parse_key(&mPKey, reinterpret_cast<const unsigned char *>(sSrvKey), sSrvKeyLength,
```
See https://github.com/Mbed-TLS/mbedtls/blob/development/docs/3.0-migration-guide.md#some-functions-gained-an-rng-parameter

edmont replied:
```
#if (MBEDTLS_VERSION_NUMBER >= 0x03000000)
#include "crypto/mbedtls.hpp"
            int rv = mbedtls_pk_parse_key(&mPKey, reinterpret_cast<const unsigned char *>(sSrvKey), sSrvKeyLength,
                                          nullptr, 0, Crypto::MbedTls::CryptoSecurePrng, nullptr);
#else
            int rv = mbedtls_pk_parse_key(&mPKey, reinterpret_cast<const unsigned char *>(sSrvKey), sSrvKeyLength,
                                          nullptr, 0);
#endif
```

samkumar replied:
Thanks. I've included this in my next commit.

edmont reviewed (commented):
Thanks Sam!

I spotted a couple of build issues when using Mbed TLS 3.x. Please see the links in the comments.

edmont commented on src/cli/cli_tcp.cpp:
```diff
@@ -74,13 +98,62 @@ template <> otError TcpExample::Process<Cmd("init")>(Arg aArgs[])
     else
     {
-        if (aArgs[0] == "circular")
+        if (aArgs[0] == "tls")
```
Update `README_TCP.md`?

edmont commented on src/core/api/tcp_ext_api.cpp:
```diff
@@ -79,4 +79,48 @@ otError otTcpCircularSendBufferDeinitialize(otTcpCircularSendBuffer *aSendBuffer
+    otError                             error;
+
+    error = otTcpCircularSendBufferWrite(endpoint, sendBuffer, aBuf, aLen, &bytes_written, 0);
+    VerifyOrExit(error == OT_ERROR_NONE, result = MBEDTLS_ERR_SSL_INTERNAL_ERROR);
```
Need this:

`#include <mbedtls/debug.h>
`

samkumar replied:
Sorry, I'm not seeing why we need this `#include`?

edmont commented:
After building from `ot-nrf528xx` I get the following errors:

```
> tcp init tls                                                                            
mbedtls_ssl_setup returned -32512                                                         
Done  
> tcp listen :: 3000                                                                      
Done                                                                                      
> Accepted connection from [fe80:0:0:0:8463:2e69:a3cf:67c2]:49152                         
TCP: Connection established                                                               
mbedtls_ssl_set_hs_ecjpake_password returned -28928                                       
TLS Handshake Failed: -28928                                                              
TLS Handshake Failed: -28928
```

edmont commented on src/cli/cli_tcp.cpp:
```diff
@@ -74,13 +98,62 @@ template <> otError TcpExample::Process<Cmd("init")>(Arg aArgs[])
+            rv = mbedtls_ssl_setup(&mSSLContext, &mSSLConfig);
+            if (rv != 0)
+            {
+                OutputLine("mbedtls_ssl_setup returned %d", rv);
```
I hit this error when using Mbed TLS 3.x, fixed it using a similar approach as in https://github.com/openthread/openthread/blob/main/src/core/meshcop/dtls.cpp#L290-L296

samkumar replied:
I included similar code (`#if (MBEDTLS_VERSION_NUMBER >= 0x03020000)`) as well here --- can you check if the fix I implemented is the same as the one that worked for you earlier?

edmont commented:
Hi @samkumar, did you have a chance to ponder my review comments?

samkumar commented:
Hi @edmont, sorry for the radio silence on this. I've been busy with job applications this past month and lost track of this. I'll try to get to this this week.

samkumar commented:
Sorry for not getting to this yet. I have another round of job applications but I think I'll get a chance after Thanksgiving in early December.

samkumar commented:
@edmont I am taking a look at this tonight and tomorrow. Is there a particular build configuration of OpenThread you used to get this to work with MbedTLS 3 or did you manually update `third_party/mbedtls/repo`?

samkumar commented:
I manually updated `third_party/mbedtls/repo` to Mbed TLS 3.3.0 (released a week ago) and verified the problems that @edmont noticed. I also incorporated the fixes. The only exception was `#include <mbedtls/debug.h>`; the code seems to work without it, but I am happy to include that anyway if others would prefer.

Tomorrow I'll update the `README.md` file as requested and look into why some of the build checks are failing.

samkumar commented:
I've updated the README and cleaned up the configuration to fix the build issues. I also squashed the commits.

edmont reviewed (commented):
Thanks @samkumar! Now it builds fine for me. However I'm having some issues testing it. Did you test with nRF52840? I had to increase the internal heap size in order to avoid `MBEDTLS_ERR_SSL_ALLOC_FAILED` after issuing `tcp init tls`.

But after that I'm getting another error:
```
> tcp connect fe80:0:0:0:146e:a00:0:1 30000                                                                             
Done                                                                                                                    
> TCP: Connection established                                                                                           
TLS Handshake Failed: -19840                                                                                            
TLS Handshake Failed: -19840
```

Adding a few initial comments in the code.

samkumar commented on src/core/config/ip6.h:
```diff
@@ -172,6 +172,16 @@
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_TLS_ENABLE
+#define OPENTHREAD_CONFIG_TLS_ENABLE 1
```
I will include this in my next update.

samkumar commented on src/cli/README_TCP.md:
```diff
@@ -157,14 +157,19 @@ stoplistening
 
 Initializes the example TCP listener and the example TCP endpoint.
 
+- mode: this specifies the buffering strategy and whether to use TLS. The possible values are "linked", "circular", and "tls".
```
Thanks. I have included this in my next update, which I will push shortly.

samkumar commented on src/cli/cli_tcp.cpp:
```diff
@@ -74,13 +102,80 @@ template <> otError TcpExample::Process<Cmd("init")>(Arg aArgs[])
+                OutputLine("mbedtls_ssl_setup returned %d", rv);
+            }
+        }
+        else if (aArgs[0] == "circular")
```
I did it this way initially, but it confuses `clang-format` and causes it to indent the `if` statement by an extra four spaces. That's why I did it as above, with the `#else` clause.

I'm happy to do it the way you suggest if you're OK with the `if` statement being indented by an extra four spaces.

edmont replied:
I'd tend to the solution with less lines of code, but not strong opinion here.

samkumar replied:
I realized that this can be cleanly resolved by changing the order of the `if`/`else if` clauses so that TLS is not the first. This way, we can avoid using `#else` without confusing `clang-format`. I included this in my latest update.

samkumar commented:
I assume that the problems you were having have to do with mbedtls3?

I went ahead and tried running the code with mbed TLS 3.3.0 on two Nordic NRF52840-DK boards, and everything worked as expected (i.e., I couldn't reproduce the errors that you were running into).

Here is a ZIP of a packet capture of it working, collected using a packet sniffer: [tls-example-capture-mbedtls3.zip](https://github.com/openthread/openthread/files/10290283/tls-example-capture-mbedtls3.zip). The decryption key is `00f7675bf5e87847f0ab6e318464c1f6` and the Mesh Local Prefix is `fdab:f2a7:5c4a:9b20::/64`.

There may have been differences between how you ran mbed TLS 3 and how I ran it. An possible next step is for you to try running it the same way I did in case that resolves the problems you were facing. I've provided my code, which includes this pull request and integrates mbed TLS 3.3.0 into OpenThread in the `feature-tcp-cli-tls-mbedtls3` branch of my fork of OpenThread: https://github.com/samkumar/openthread/tree/feature-tcp-cli-tls-mbedtls3. To run it, first go to `third_party/mbedtls/repo` and run `scripts/generate_ssl_debug_helpers.py`; this will generate `third_party/mbedtls/repo/library/ssl_debug_helpers_generated.c`. Then build OpenThread as usual.

samkumar commented:
@edmont I just rebased the code on the latest `main` branch, and checked that mbed TLS 3.3.0 still works on the NRF52840-DK boards.

samkumar commented:
In case it helps, here are the commands I ran on the two boards: 
[tls-example-capture-mbedtls3-notes.txt](https://github.com/openthread/openthread/files/10290505/tls-example-capture-mbedtls3-notes.txt).

edmont commented:
Sorry I didn't mention before, I'm not using TLS 3.3.0 for the initial testing, just checked out your branch on top of https://github.com/openthread/ot-nrf528xx, built and flashed as described in https://github.com/openthread/ot-nrf528xx/blob/main/src/nrf52840/README.md#building-the-examples

After today's updates I'm still having the same issues I described yesterday.

samkumar commented:
Here is the `.hex` file I've been programming my devices with using `nrfjprog`: [ot-cli-ftd.hex.zip](https://github.com/openthread/openthread/files/10296390/ot-cli-ftd.hex.zip). You'll need to `unzip` it first. Can you try using this? This will let us narrow down whether it is due to a difference in how we are building the code or whether it's a hardware difference between the boards you're using and boards I'm using.

samkumar commented:
I also looked more closely at the errors you are seeing, and it seems that they are related to memory capacity: `MBEDTLS_ERR_SSL_ALLOC_FAILED` and `MBEDTLS_ERR_ECP_ALLOC_FAILED`.

I realized that the reason I'm not seeing these errors might be that I am building the code with `-DOT_FULL_LOGS=OFF`. The full build command I am using is `script/build nrf52840 UART_trans -DOT_JOINER=ON -DOT_COMMISSIONER=ON -DOT_FULL_LOGS=OFF`. `-DOT_FULL_LOGS=OFF` is needed to get good TCP performance, which is why I've always been building the code that way. I managed to reproduce the `MBEDTLS_ERR_SSL_ALLOC_FAILED` error you were seeing by building without `-DOT_FULL_LOGS=OFF`. It seems that, when full logging is enabled, there is not enough leftover memory for TLS, resulting in memory allocation issues.

I noticed that OpenThread solves this problem for DTLS by defining two heap sizes, `OPENTHREAD_CONFIG_HEAP_INTERNAL_SIZE` (large) and `OPENTHREAD_CONFIG_HEAP_INTERNAL_SIZE_NO_DTLS` (small), and selecting the larger one if DTLS is enabled. I updated the logic in `src\core\utils\heap.hpp` to use the larger heap size if TLS is enabled (even if DTLS is disabled). With this fix, I managed to run TLS on the nrf52840-DK even when building with just `script/build nrf52840 UART_trans`. I just updated the pull request with this change. Can you check if it resolves the issues you were seeing?

edmont reviewed (approved):
Thanks! That was the issue, now it works fine for me.

samkumar commented:
Thanks @edmont for reviewing this!

@jwhui Can you please take a look and merge this pull request if it looks OK? I just rebased it onto the latest `main` branch.

jwhui commented on src/cli/cli_tcp.cpp:
```diff
@@ -55,10 +60,33 @@ TcpExample::TcpExample(otInstance *aInstance, OutputImplementer &aOutputImplemen
+{
+    OT_UNUSED_VARIABLE(ctx);
+    otError error = ot::Random::Crypto::FillBuffer(static_cast<uint8_t *>(buffer), static_cast<uint16_t>(length));
+    if (error != OT_ERROR_NONE)
+    {
+        return -1;
+    }
+    return 0;
```
nit: prefer a single return statement
```suggestion
    return (error == OT_ERROR_NONE) ? 0 : -1;
```

samkumar replied:
I fixed this in the latest commit.

jwhui commented on src/cli/cli_tcp.hpp:
```diff
@@ -108,13 +121,19 @@ class TcpExample : private Output
     bool mEndpointConnected;
     bool mSendBusy;
     bool mUseCircularSendBuffer;
+    bool mUseTLS;
+    bool mTLSHandshakeComplete;
```
nit: prefer camel case acronyms
```suggestion
    bool mUseTls;
    bool mTlsHandshakeComplete;
```

samkumar replied:
I fixed this in the latest commit.

jwhui commented on src/cli/cli_tcp.hpp:
```diff
@@ -127,8 +146,71 @@ class TcpExample : private Output
+    otTcpEndpointAndCircularSendBuffer mEndpointAndCircularSendBuffer;
+
+#if OPENTHREAD_CONFIG_TLS_ENABLE
+    mbedtls_ssl_context      mSSLContext;
+    mbedtls_ssl_config       mSSLConfig;
```
nit:
```suggestion
    mbedtls_ssl_context      mSslContext;
    mbedtls_ssl_config       mSslConfig;
```

samkumar replied:
I fixed this in the latest commit.

jwhui commented on include/openthread/tcp_ext.h:
```diff
@@ -215,6 +215,35 @@ void otTcpCircularSendBufferForceDiscardAll(otTcpCircularSendBuffer *aSendBuffer
+ *
+ * @param[in]  aCtx  A pointer to an otTcpEndpointAndCircularSendBuffer.
+ * @param[in]  aBuf  The data to add to the send buffer.
+ * @param[in]  aLen  The amount of data to add to the send buffer.
+ * @return The number of bytes sent, or an mbedtls error code.
```
nit:
```suggestion
 * @param[in]  aLen  The amount of data to add to the send buffer.
 *
 * @returns The number of bytes sent, or an mbedtls error code.
```

samkumar replied:
I fixed this in the latest commit.

jwhui commented on include/openthread/tcp_ext.h:
```diff
@@ -215,6 +215,35 @@ void otTcpCircularSendBufferForceDiscardAll(otTcpCircularSendBuffer *aSendBuffer
+ *
+ * @param[in]   aCtx  A pointer to an otTcpEndpointAndCircularSendBuffer.
+ * @param[out]  aBuf  The buffer into which to receive data.
+ * @param[in]   aLen  The maxmimum amount of data that can be received.
+ * @return The number of bytes received, or an mbedtls error code.
```
nit:
```suggestion
 * @param[in]   aLen  The maxmimum amount of data that can be received.
 *
 * @returns The number of bytes received, or an mbedtls error code.
```

samkumar replied:
I fixed this in the latest commit. I also used `@returns` instead of `@return` throughout the file to be consistent.

jwhui reviewed (commented):
Looks great, thanks! Just a few very minor nits.

samkumar commented:
@jwhui I made the changes you requested and rebased onto the latest `main` branch. Sorry for the delay.

jwhui reviewed (approved):
Looks great. Thanks @samkumar ! 👍🏼
